compiler/rustc_serialize/src/opaque.rs

```diff
@@ -38,210 +35,97 @@ pub struct FileEncoder {
-                ptr::copy_nonoverlapping(src, dst, buf_len);
-            }
-
-            self.buffered = buffered + buf_len;
+        if let Some(dest) = self.buffer_empty().get_mut(..buf.len()) {
+            dest.copy_from_slice(buf);
+            self.buffered += buf.len();
         } else {
-            self.write_all_unbuffered(buf);
+            self.write_all_cold_path(buf);
         }
     }
 
-    fn write_all_unbuffered(&mut self, mut buf: &[u8]) {
-        // If we've already had an error, do nothing. It'll get reported after
-        // `finish` is called.
-        if self.res.is_err() {
-            return;
-        }
-
-        if self.buffered > 0 {
+    /// Write up to `N` bytes to this encoder.
+    ///
+    /// Whenever possible, use this function to do writes whose length has a small and
+    /// compile-time constant upper bound.
+    #[inline]
+    pub fn write_with<const N: usize, V>(&mut self, mut visitor: V)
+    where
+        V: FnMut(&mut [u8; N]) -> usize,
+    {
+        let flush_threshold = const { BUF_SIZE.checked_sub(N).unwrap() };
+        if std::intrinsics::unlikely(self.buffered > flush_threshold) {
             self.flush();
         }
-
-        // This is basically a copy of `Write::write_all` but also updates our
-        // `self.flushed`. It's necessary because `Write::write_all` does not
-        // return the number of bytes written when an error is encountered, and
-        // without that, we cannot accurately update `self.flushed` on error.
-        while !buf.is_empty() {
-            match self.file.write(buf) {
-                Ok(0) => {
-                    self.res = Err(io::Error::new(
-                        io::ErrorKind::WriteZero,
-                        "failed to write whole buffer",
-                    ));
-                    return;
-                }
-                Ok(n) => {
-                    buf = &buf[n..];
-                    self.flushed += n;
-                }
-                Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {}
-                Err(e) => {
-                    self.res = Err(e);
-                    return;
-                }
-            }
-        }
+        // SAFETY: We checked above that that N < self.buffer_empty().len(),
+        // and if isn't, flush ensures that our empty buffer is now BUF_SIZE.
+        // We produce a post-mono error if N > BUF_SIZE.
+        let buf = unsafe { self.buffer_empty().first_chunk_mut::<N>().unwrap_unchecked() };
+        let written = visitor(buf);
+        debug_assert!(written <= N);
+        // We have to ensure that an errant visitor cannot cause self.buffered to exeed BUF_SIZE.
+        self.buffered += written.min(N);
     }
 
     pub fn finish(mut self) -> Result<usize, io::Error> {
         self.flush();
-
-        let res = std::mem::replace(&mut self.res, Ok(()));
-        res.map(|()| self.position())
-    }
-}
-
-impl Drop for FileEncoder {
```

Comment:
I'm actually not sure what the right thing to do here is. I can add back the `Drop` impl, but if we have encountered an error there is nothing useful to do with it. We can't report it because we don't have `TyCtxt` in this crate. We can't return it, and panicking with the error if there is one would probably just produce rare ICE reports.

I suppose we could make the `Drop` impl just an `unreachable!` that reminds the user to call `finish`?

Reply:
For now I've put back the `Drop` impl. That seems the least controversial if nothing else?